[serialization] no transitive decl change

ChuanqiXu9 opened the pull request:
Following of https://github.com/llvm/llvm-project/pull/86912

#### Motivation Example

The motivation of the patch series is that, for a module interface unit `X`, when the dependent modules of `X` changes, if the changes is not relevant with `X`, we hope the BMI of `X` won't change. For the specific patch, we hope if the changes was about irrelevant declaration changes, we hope the BMI of `X` won't change. **However**, I found the patch itself is not very useful in practice, since the adding or removing declarations, will change the state of identifiers and types in most cases.

That said, for the most simple example,

```
// partA.cppm
export module m:partA;

// partA.v1.cppm
export module m:partA;
export void a() {}

// partB.cppm
export module m:partB;
export void b() {}

// m.cppm
export module m;
export import :partA;
export import :partB;

// onlyUseB;
export module onlyUseB;
import m;
export inline void onluUseB() {
    b();
}
```

the BMI of `onlyUseB` will change after we change the implementation of `partA.cppm` to `partA.v1.cppm`. Since `partA.v1.cppm` introduces new identifiers and types (the function prototype).

So in this patch, we have to write the tests as:

```
// partA.cppm
export module m:partA;
export int getA() { ... }
export int getA2(int) { ... }

// partA.v1.cppm
export module m:partA;
export int getA() { ... }
export int getA(int) { ... }
export int getA2(int) { ... }

// partB.cppm
export module m:partB;
export void b() {}

// m.cppm
export module m;
export import :partA;
export import :partB;

// onlyUseB;
export module onlyUseB;
import m;
export inline void onluUseB() {
    b();
}
```

so that the new introduced declaration `int getA(int)` doesn't introduce new identifiers and types, then the BMI of `onlyUseB` can keep unchanged.

While it looks not so great, the patch should be the base of the patch to erase the transitive change for identifiers and types since I don't know how can we introduce new types and identifiers without introducing new declarations. Given how tightly the relationship between declarations, types and identifiers, I think we can only reach the ideal state after we made the series for all of the three entties.

#### Design details

The design of the patch is similar to https://github.com/llvm/llvm-project/pull/86912, which extends the 32-bit DeclID to 64-bit and use the higher bits to store the module file index and the lower bits to store the Local Decl ID. 

A slight difference is that we only use 48 bits to store the new DeclID since we try to use the higher 16 bits to store the module ID in the prefix of Decl class. Previously, we use 32 bits to store the module ID and 32 bits to store the DeclID. I don't want to allocate additional space so I tried to make the additional space the same as 64 bits. An potential interesting thing here is about the relationship between the module ID and the module file index. I feel we can get the module file index by the module ID. But I didn't prove it or implement it. Since I want to make the patch itself as small as possible. We can make it in the future if we want.

Another change in the patch is the new concept Decl Index, which means the index of the very big array `DeclsLoaded` in ASTReader. Previously, the index of a loaded declaration is simply the Decl ID minus PREDEFINED_DECL_NUMs. So there are some places they got used ambiguously. But this patch tried to split these two concepts.

#### Overhead

As https://github.com/llvm/llvm-project/pull/86912 did, the change will increase the on-disk PCM file sizes. As the declaration ID may be the most IDs in the PCM file, this can have the biggest impact on the size. In my experiments, this change will bring 6.6% increase of the on-disk PCM size. No compile-time performance regression observed. Given the benefits in the motivation example, I think the cost is worthwhile.

ChuanqiXu9 committed:
[serialization] no transitive decl change

ChuanqiXu9 commented:
@jansvoboda11 @Bigcheese gentle ping

jansvoboda11 commented on clang/include/clang/AST/DeclID.h:
```diff
@@ -124,6 +130,15 @@ class DeclIDBase {
+  unsigned getModuleFileIndex() const { return ID >> 32; }
+
+  unsigned getLocalDeclIndex() const {
+    // Implement it directly instead of calling `llvm::maskTrailingOnes` since
+    // we don't want `MathExtras.h` to be inclued here.
```
Why not? Just to keep the amount of code lower? Can't we move the function body to an implementation file?

ChuanqiXu9 replied:
Maybe I just didn't want to create an implementation file for such a single simple function. But I just realized I can append it in `DeclBase.cpp` just like many other classes did. So done.

jansvoboda11 commented on clang/include/clang/Serialization/ASTBitCodes.h:
```diff
@@ -255,6 +255,12 @@ class DeclOffset {
 };
 
+// The unaligned decl ID used in the Blobs of bistreams.
+using unalighed_decl_id_t =
```
I know this typo isn't introduced by your PR, but since your PR is touching it, please fix it in a prep commit.
```suggestion
using unaligned_decl_id_t =
```

ChuanqiXu9 replied:
Nice catch! Done in https://github.com/llvm/llvm-project/commit/e73e4951b20c70f24354e2a2820876c818dcaee3

jansvoboda11 commented on clang/lib/Serialization/ASTReader.cpp:
```diff
@@ -7802,20 +7800,31 @@ Decl *ASTReader::GetDecl(GlobalDeclID ID) {
-  assert(I != GlobalDeclMap.end() && "Corrupted global declaration map");
-  ModuleFile *Owner = I->second;
+  uint64_t OrignalModuleFileIndex = 0;
+  for (unsigned I = 0; I < M.DependentModules.size(); I++)
```
I just now noticed the name `DependentModules` (introduced in #86912) sounds like exactly the opposite of what this member represents. Can we rename this to `TransitiveImports` (or something similar) in a prep patch to clarify the semantics here?

ChuanqiXu9 replied:
Done in https://github.com/llvm/llvm-project/commit/b590ba73a76609bace9949ea8195d2ee8213cb3f

jansvoboda11 reviewed (commented):
This looks good conceptually, I left a couple of minor notes.

ChuanqiXu9 commented on clang/lib/AST/DeclBase.cpp:
```diff
@@ -111,6 +109,28 @@ void *Decl::operator new(std::size_t Size, const ASTContext &Ctx,
+GlobalDeclID Decl::getGlobalID() const {
+  if (!isFromASTFile())
+    return GlobalDeclID();
+  uint64_t ID = *((const uint64_t *)this - 1);
```
Previously, for deserialized declaration, clang would apply an additional 64 bits for it. And clang would store the global declaration ID at the lower 32 bits and store the module index at the higher 32 bits.

And after this patch, we extend the global declaration ID to 64 bits and I don't want to apply additional spaces. So I decided to use the lower 48 bits to store the declaration ID and the higher 16 bits to store the module index.

I added a comment here to clarify that.

And as I mentioned in the message, it looks like there are relationships between module index and module file index (the upper 16 bits of the new declaration ID). But on the one hand, I want to make the patch as simple as possible to make it land faster. On the other hand, it looks like 16 bits is enough for both module index and module file index. And we can't get any improvements by merge it. So I just leave it as is.

jansvoboda11 replied:
Makes sense, thanks!

ChuanqiXu9 commented:
Thanks for reviewing : )

ChuanqiXu9 commented:
I'd like to land this since:
- I want to give more time testing this
- the design of the patch is pretty similar with the previous one (https://github.com/llvm/llvm-project/pull/86912)
- the scale of the patch is not very big (100+ lines of code except new tests)
- I do think the functionality is very very important to modules.

ChuanqiXu9 commented:
I'll close it as there are some crashes.

---

The crash log:

```
********** TEST 'Clang :: Modules/redecl-ivars.m' FAILED ********************
Exit Code: 1

Command Output (stderr):
--
RUN: at line 2: rm -rf /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/tools/clang/test/Modules/Output/redecl-ivars.m.tmp
+ rm -rf /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/tools/clang/test/Modules/Output/redecl-ivars.m.tmp
RUN: at line 3: split-file /b/sanitizer-x86_64-linux-fast/build/llvm-project/clang/test/Modules/redecl-ivars.m /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/tools/clang/test/Modules/Output/redecl-ivars.m.tmp
+ split-file /b/sanitizer-x86_64-linux-fast/build/llvm-project/clang/test/Modules/redecl-ivars.m /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/tools/clang/test/Modules/Output/redecl-ivars.m.tmp
RUN: at line 4: /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/clang -cc1 -internal-isystem /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/lib/clang/19/include -nostdsysteminc -fsyntax-only -fobjc-runtime=macosx-10.9 -verify -I/b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/tools/clang/test/Modules/Output/redecl-ivars.m.tmp/include /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/tools/clang/test/Modules/Output/redecl-ivars.m.tmp/test-mismatch-in-extension.m
+ /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/clang -cc1 -internal-isystem /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/lib/clang/19/include -nostdsysteminc -fsyntax-only -fobjc-runtime=macosx-10.9 -verify -I/b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/tools/clang/test/Modules/Output/redecl-ivars.m.tmp/include /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/tools/clang/test/Modules/Output/redecl-ivars.m.tmp/test-mismatch-in-extension.m
RUN: at line 5: /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/clang -cc1 -internal-isystem /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/lib/clang/19/include -nostdsysteminc -fsyntax-only -fobjc-runtime=macosx-10.9 -verify -I/b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/tools/clang/test/Modules/Output/redecl-ivars.m.tmp/include /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/tools/clang/test/Modules/Output/redecl-ivars.m.tmp/test-mismatch-in-extension.m             -fmodules -fimplicit-module-maps -fmodules-cache-path=/b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/tools/clang/test/Modules/Output/redecl-ivars.m.tmp/modules.cache
+ /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/bin/clang -cc1 -internal-isystem /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/lib/clang/19/include -nostdsysteminc -fsyntax-only -fobjc-runtime=macosx-10.9 -verify -I/b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/tools/clang/test/Modules/Output/redecl-ivars.m.tmp/include /b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/tools/clang/test/Modules/Output/redecl-ivars.m.tmp/test-mismatch-in-extension.m -fmodules -fimplicit-module-maps -fmodules-cache-path=/b/sanitizer-x86_64-linux-fast/build/llvm_build_asan_ubsan/tools/clang/test/Modules/Output/redecl-ivars.m.tmp/modules.cache
/b/sanitizer-x86_64-linux-fast/build/libcxx_build_asan_ubsan/include/c++/v1/__algorithm/lower_bound.h:39:53: runtime error: reference binding to misaligned address 0x529000045c44 for type 'const clang::serialization::ObjCCategoriesInfo', which requires 8 byte alignment
0x529000045c44: note: pointer points here
  00 00 00 00 02 00 00 00  01 00 00 00 02 00 00 00  00 00 00 00 c3 0d 88 60  0a a8 81 10 03 20 08 00
              ^ 
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /b/sanitizer-x86_64-linux-fast/build/libcxx_build_asan_ubsan/include/c++/v1/__algorithm/lower_bound.h:39:53 

--

It looks like misaligned access too. I'll try to fix it.

(It is somewhat hurting that we can't find them on the trunk.)
```

hctim commented:
Hi from the sanitizer buildbot mainter, thanks for the fast revert! For full reproduction, you can use https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild and `buildbot_fast.sh`. But, you can reproduce it quicker by using `-DCMAKE_C_FLAGS="-fsanitize=undefined"  -DCMAKE_CXX_FLAGS="-fsanitize=undefined" -DLLVM_USE_SANITIZER=Undefined` in your cmake (then `{ninja,make} check-clang`).

Michael137 commented:
I.e., also broke the macOS LLDB buildbots: https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/lldb-cmake/2637/execution/node/97/log/

```
Assertion failed: (!((*IDAddress) >> 48) && "We should only set the module ID once"), function setOwningModuleID, file DeclBase.cpp, line 132.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
```

ChuanqiXu9 commented:
Thanks for reporting. I've reproduced them (including the lldb test) locally and fixed them. I'll try to land it again to see what happens.

Michael137 commented:
Unfortunately this is still failing one test on the macOS CI: https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/5083/execution/node/97/log

```
======================================================================
FAIL: test_duplicate_decls_gmodules (TestTemplateWithSameArg.TestTemplateWithSameArg)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/packages/Python/lldbsuite/test/lldbtest.py", line 1756, in test_method
    return attrvalue(self)
  File "/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/test/API/lang/cpp/gmodules/template-with-same-arg/TestTemplateWithSameArg.py", line 66, in test_duplicate_decls
    self.filecheck("target module dump ast", __file__)
  File "/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/packages/Python/lldbsuite/test/lldbtest.py", line 2293, in filecheck
    self.assertTrue(cmd_status == 0)
AssertionError: False is not true
Config=arm64-/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/lldb-build/bin/clang
----------------------------------------------------------------------
Ran 2 tests in 1.682s

FAILED (failures=1)
```

Let me know if you need help reproducing this.

Looks like we're doing a FileCheck on the imported LLDB AST:
```
FileCheck output:

/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/test/API/lang/cpp/gmodules/template-with-same-arg/TestTemplateWithSameArg.py:69:10: error: CHECK: expected string not found in input
# CHECK: ClassTemplateSpecializationDecl {{.*}} imported in Module2 struct ClassInMod3 definition
```

But instead of `imported in Module2 struct ClassInMod3 definition` it's now just `imported struct ClassInMod3 definition`

ChuanqiXu9 commented:
Oh, this is not wanted. It should be almost a NFC patch for users. I'll revert it.

ChuanqiXu9 commented:
> Unfortunately this is still failing one test on the LLDB macOS CI: https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/5083/execution/node/97/log
> 
> ```
> ======================================================================
> FAIL: test_duplicate_decls_gmodules (TestTemplateWithSameArg.TestTemplateWithSameArg)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/packages/Python/lldbsuite/test/lldbtest.py", line 1756, in test_method
>     return attrvalue(self)
>   File "/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/test/API/lang/cpp/gmodules/template-with-same-arg/TestTemplateWithSameArg.py", line 66, in test_duplicate_decls
>     self.filecheck("target module dump ast", __file__)
>   File "/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/packages/Python/lldbsuite/test/lldbtest.py", line 2293, in filecheck
>     self.assertTrue(cmd_status == 0)
> AssertionError: False is not true
> Config=arm64-/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/lldb-build/bin/clang
> ----------------------------------------------------------------------
> Ran 2 tests in 1.682s
> 
> FAILED (failures=1)
> ```
> 
> Let me know if you need help reproducing this.
> 
> Looks like we're doing a FileCheck on the imported LLDB AST:
> 
> ```
> FileCheck output:
> 
> /Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/test/API/lang/cpp/gmodules/template-with-same-arg/TestTemplateWithSameArg.py:69:10: error: CHECK: expected string not found in input
> # CHECK: ClassTemplateSpecializationDecl {{.*}} imported in Module2 struct ClassInMod3 definition
> ```
> 
> But instead of `imported in Module2 struct ClassInMod3 definition` it's now just `imported struct ClassInMod3 definition`. Maybe we're not setting the module ID correctly in LLDB or something? Does that ring any bells?

@Michael137 hi, how can I run test for it? I tried

```
bin/llvm-lit ../lldb/test/API/lang/cpp/gmodules/template-with-same-arg/TestTemplateWithSameArg.py
```

but it shows it can't work

DavidSpickett commented:
Maybe you can run individual tests with lit but generally I use `lldb-dotest` instead:
```
./bin/lldb-dotest.py -p TestTemplateWithSameArg.py
```
(you only need the filename)

ChuanqiXu9 commented:
> Maybe you can run individual tests with lit but generally I use `lldb-dotest` instead:
> 
> ```
> ./bin/lldb-dotest.py -p TestTemplateWithSameArg.py
> ```
> 
> (you only need the filename)

Thanks. Reproduced. But it surprised me that I can't run the commands you mentioned, but I need to run:

```
./bin/lldb-dotest -p TestTemplateWithSameArg.py -G gmodules
```

And I am also slightly surprised that after I change the code, it doesn't work if I run `ninja lldb-test` only I need to run `ninja clang lldb`. Do I misconfigure anything? I build lldb by:

```
CC=clang CXX=clang++ cmake -DLLVM_ENABLE_PROJECTS="clang;lldb"  -GNinja ../llvm -DCMAKE_BUILD_TYPE=Release  -DLLVM_ENABLE_ASSERTIONS=On -DLLVM_ENABLE_LLD=ON  -DPython3_ROOT_DIR=<path-to-python>
```

ChuanqiXu9 commented:
I've fixed the lldb failure and resent https://github.com/llvm/llvm-project/commit/97c866f6c86456b3316006e6beff47e68a81c00a.

DavidSpickett commented:
> Thanks. Reproduced. But it surprised me that I can't run the commands you mentioned, but I need to run:
> ./bin/lldb-dotest -p TestTemplateWithSameArg.py -G gmodules

Running the tests as modules is perhaps opt in, we have the ability to run them with different debug info types too. So this is not surprising. Well, it is for you, but doesn't mean there's anything wrong here :)

> And I am also slightly surprised that after I change the code, it doesn't work if I run ninja lldb-test only I need to run ninja clang lldb. Do I misconfigure anything? I build lldb by:

My comment about `lld-test` referred to a different failing test I think (and if we did have some dependencies incorrect, it wouldn't surprise me). Nothing wrong with that config line though.

So good news is Linux and Windows lldb are ok (Mac will report back soon I assume), but now we have failures on 32 bit Arm builds:
https://lab.llvm.org/buildbot/#/builders/245/builds/25498

I'm going to reproduce this shortly but my bet is that this is undefined behaviour of some kind. It may not be due to your code but your code may be exposing a pre-existing issue.

DavidSpickett commented:
> I didn't recognize we may still have 32-bit machines

I think 32 bit Arm are the last.

>  I'll revert it again.

Sure, if you post the PR I can test the fix too. I know getting 32 bit machines can be tricky.

ChuanqiXu9 commented:
> > I didn't recognize we may still have 32-bit machines
> 
> I think 32 bit Arm are the last.
> 
> > I'll revert it again.
> 
> Sure, if you post the PR I can test the fix too. I know getting 32 bit machines can be tricky.

I sent https://github.com/llvm/llvm-project/pull/94603 . I feel it should at least be one of the problem but I am not sure if it can fix all the issues.

ChuanqiXu9 commented:
Resent https://github.com/llvm/llvm-project/commit/5c104879c1a98eeb845c03e7c45206bd48e88f0c

Thanks for testing it !

ChuanqiXu9 commented:
Sadly it looks like the (refined) patch still breaks the alignment on 32 bit machine. https://lab.llvm.org/buildbot/#/builders/174/builds/34307

I've reverted the patch and updated https://github.com/llvm/llvm-project/pull/94603 to test it again.

ChuanqiXu9 commented:
Thanks for testing. I've resent https://github.com/llvm/llvm-project/commit/5a0181f568e56e37df80d0f74eca4775776fa8cd.

joanahalili commented:
just a heads up: we are seeing a number targets failing due to an compilation timeout caused by this commit.

asmok-g commented:
> just a heads up: we are seeing a number targets failing due to a compilation timeout caused by this commit.

More precisely, the timeouts are happening only on arm CPUs.

ChuanqiXu9 commented:
Oh, this is surprising. Are arm CPUs 32 bit? And how much do these tests get slowed down?

alexfh commented:
The report is about compilation timeouts rather than runtime performance. The code being compiled is outside the tree. We're working on a shareable reproducer.

alexfh commented:
As the architecture is concerned, it's AArch64, which is 64 bit.

bgra8 commented:
@ChuanqiXu9 the description of the change says that the maximum expected PCMs size increase is 6.6%. In our testing at google we see noticed for some large targets the total combined PCM size increase by more than 25% (both ARM64 and x86_64).

alexfh commented:
So, what's happening here is a significant increase in the compilation time of a pretty large source that has a large number of modular dependencies (using Clang header modules rather than C++20 modules).

The times reported by `clang -ftime-report` for clang frontend change drastically:

before this commit:
```
===-------------------------------------------------------------------------===
                          Clang front-end time report
===-------------------------------------------------------------------------===
  Total Execution Time: 207.1765 seconds (218.7714 wall clock)

   ---User Time---   --System Time--   --User+System--   ---Wall Time---  --- Name ---
  153.2751 ( 78.6%)   8.5109 ( 70.4%)  161.7860 ( 78.1%)  170.5710 ( 78.0%)  Clang front-end timer
  40.0169 ( 20.5%)   3.1469 ( 26.0%)  43.1638 ( 20.8%)  45.8168 ( 20.9%)  Reading modules
```

after this commit:
```
===-------------------------------------------------------------------------===
                          Clang front-end time report
===-------------------------------------------------------------------------===
  Total Execution Time: 1134.1241 seconds (1179.8429 wall clock)
                                                                                                                                                                                                                                                        ---User Time---   --System Time--   --User+System--   ---Wall Time---  --- Name ---
  591.7280 ( 54.0%)  21.5861 ( 56.0%)  613.3141 ( 54.1%)  638.4518 ( 54.1%)  Clang front-end timer
  502.9522 ( 45.9%)  16.6865 ( 43.3%)  519.6387 ( 45.8%)  540.1083 ( 45.8%)  Reading modules
```

Given the notorious difficulty in preparing shareable test cases for modular compilations, I expect this to take days until we can give you something specific, but I'll ask @ilya-biryukov if he can spot something right away.

And I suspect that ARM is a red herring here. The problem seems to be purely in Clang frontend. Checking...

ChuanqiXu9 commented:
Oh, the time regression is surprising to me. And I observed that @alexfh only reports (compile performance) problems on AArch64 but @bgra8 reports (size increase problems) on both ARM64 and X86_64. 

@alexfh is it true?

So I **guess** the problem may come from the unaligned access. I'll try to refactor it to store 2 32-bits integer instead of 1 64 bits in the PCM files. It may still be helpful to locate the problem and prepare a reproducer though.

ChuanqiXu9 commented:
I sent https://github.com/llvm/llvm-project/pull/95348 for aligned related change.

For size increase, the reason and the possible solution are clear. We should store the declaration ID as two slots instead of one in the serialized format so that we can utilize VBR6 format better.

ilya-biryukov commented:
The 10x increase in the time spent reading modules does look surprising and I would need to check if it's related to PCMs becoming larger or is localized to that particular compile. (Even if on a single example).

@ChuanqiXu9 do you expect alignment to have such a drastic effect on deserialization time? That would be very surprising to me.
The reproducer will take time, could we revert in the meantime?

ChuanqiXu9 commented:
> The 10x increase in the time spent reading modules does look surprising and I would need to check if it's related to PCMs becoming larger or is localized to that particular compile. (Even if on a single example).
 
It should not related to the size of the PCMs otherwise we will see the regression on X86 too. Another point may be, the proportion is not correct, 25% size increase shouldn't introduce such a big change. 

BTW, I am working on a patch to mitigate the size issue. And for the different size increase number, it should be related to the specific code bases. Generally, the more references to decls, the more increases.

> 
> @ChuanqiXu9 do you expect alignment to have such a drastic effect on deserialization time? That would be very surprising to me. The reproducer will take time, could we revert in the meantime?

I won't say 'expected'. I just met some performance regression in AArch64 related to alignments. So I am just guessing it is related to alignments.

And for this particular issue, besides reproducer, it should be pretty helpful to have some hotspots data. It might not take too long.

For reverting, how about giving it a few days to locate the problem?

ilya-biryukov commented:
To clarify: we observe this regression on x86, not on Arm. @alexfh is now trying to profile the compiler to locate what's going wrong and I will try to poke at PCM files and come up with a reproducer.

However, the regressions is definitely real and will block our internal releases, preventing testing of further commits that go in (and these things may pile up). Local revert could be an option, but I'm worried that follow up changes will have conflicts (changing DeclID to 64bit seems like it should, at least potentially, have wide-reaching implications).

It is also hard to predict how long it would take to get to the bottom of it, we will still prioritize finding the problem and the reproducer if this commit is reverted, but we won't have to worry about the conflicts and such.

I do fully acknowledge that it's frustrating when problems are not found during the review or shortly after commits land, and that not having this upstream may make follow up work more difficult (because of the need to have those commits on a branch and rebase/merge more).

We try hard to be prompt with out testing, but sometimes things aren't as fast as we'd like.

alexfh commented:
To clarify a bit more: the host platform Clang runs on is x86-64 in all cases described here by myself, @joanahalili, @asmok-g, @bgra8 and @ilya-biryukov. What may be specific to aarch64 is a set of headers conditionally-included just when compiling for ARM platforms.

As @ilya-biryukov mentioned, I'm now gathering profiles from Clang performing the problematic compilation before and after this patch.

ChuanqiXu9 commented:
Yeah, it looks like there are some miscommunications. @alexfh have you tried https://github.com/llvm/llvm-project/pull/95348? This is not a big patch and I took several looks, I don't find many doubtful points. I think it is worth to give the above PR a try.

For reverting, it is really sad that the patch gets reverted without understanding what's happened and what I can do to reland it except waiting. Given you estimate we need a few days to get a shareable reproducer  and it may be an option to revert this internally, you can revert this internally and bring a reproducer back, and we can revert this and we'll understand things better.

alexfh commented:
Profile before this commit:
```
-   86.29%     0.00%  clang    clang               [.] cc1_main(llvm::ArrayRef<char const*>, char const*, void*)                                                                                                                                    ▒   - cc1_main(llvm::ArrayRef<char const*>, char const*, void*)                                                                                                                                                                                      ▒
      - 86.29% clang::ExecuteCompilerInvocation(clang::CompilerInstance*)                                                                                                                                                                           ▒
         - 86.28% clang::CompilerInstance::ExecuteAction(clang::FrontendAction&)                                                                                                                                                                    ▒
            - 83.59% clang::FrontendAction::Execute()                                                                                                                                                                                               ▒
               - 83.59% clang::ParseAST(clang::Sema&, bool, bool)                                                                                                                                                                                   ▒
                  - 69.80% clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, clang::Sema::ModuleImportState&)                                                                                                                ▒
                     - 36.99% clang::Parser::ParseExternalDeclaration(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*)                                                                                                  ▒
                        - 35.56% clang::Parser::ParseDeclaration(clang::DeclaratorContext, clang::SourceLocation&, clang::ParsedAttributes&, clang::ParsedAttributes&, clang::SourceLocation*)                                                      ▒
                           - 34.82% clang::Parser::ParseNamespace(clang::DeclaratorContext, clang::SourceLocation&, clang::SourceLocation)                                                                                                          ▒
                              - 34.67% clang::Parser::ParseInnerNamespace(llvm::SmallVector<clang::Parser::InnerNamespaceInfo, 4u> const&, unsigned int, clang::SourceLocation&, clang::ParsedAttributes&, clang::BalancedDelimiterTracker&)        ▒
                                 - 22.24% clang::Parser::ParseExternalDeclaration(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*)                                                                                      ▒
                                    - 13.58% clang::Parser::ParseDeclaration(clang::DeclaratorContext, clang::SourceLocation&, clang::ParsedAttributes&, clang::ParsedAttributes&, clang::SourceLocation*)                                          ▒
                                       - 11.95% clang::Parser::ParseNamespace(clang::DeclaratorContext, clang::SourceLocation&, clang::SourceLocation)                                                                                              ▒
                                          - 11.93% clang::Parser::ParseInnerNamespace(llvm::SmallVector<clang::Parser::InnerNamespaceInfo, 4u> const&, unsigned int, clang::SourceLocation&, clang::ParsedAttributes&, clang::BalancedDelimiterTrack▒
                                             - 11.87% clang::Parser::ParseExternalDeclaration(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*)                                                                          ▒
                                                + 7.11% clang::Parser::ParseDeclaration(clang::DeclaratorContext, clang::SourceLocation&, clang::ParsedAttributes&, clang::ParsedAttributes&, clang::SourceLocation*)                               ▒
                                                + 4.75% clang::Parser::ParseDeclarationOrFunctionDefinition(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*, clang::AccessSpecifier)                                    ▒
                                       + 1.58% clang::Parser::ParseDeclarationStartingWithTemplate(clang::DeclaratorContext, clang::SourceLocation&, clang::ParsedAttributes&)                                                                      ▒
                                    + 8.65% clang::Parser::ParseDeclarationOrFunctionDefinition(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*, clang::AccessSpecifier)                                                ▒
                                 - 7.88% clang::Parser::ParseInnerNamespace(llvm::SmallVector<clang::Parser::InnerNamespaceInfo, 4u> const&, unsigned int, clang::SourceLocation&, clang::ParsedAttributes&, clang::BalancedDelimiterTracker&)      ▒
                                    + 7.33% clang::Parser::ParseInnerNamespace(llvm::SmallVector<clang::Parser::InnerNamespaceInfo, 4u> const&, unsigned int, clang::SourceLocation&, clang::ParsedAttributes&, clang::BalancedDelimiterTracker&)   ▒
                                 + 4.52% clang::BalancedDelimiterTracker::consumeClose()                                                                                                                                                            ▒
                        + 1.42% clang::Parser::ParseDeclarationOrFunctionDefinition(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*, clang::AccessSpecifier)                                                            ▒
                     - 27.33% clang::Sema::ActOnEndOfTranslationUnit()                                                                                                                                                                              ▒
                        - 27.31% clang::Sema::ActOnEndOfTranslationUnitFragment(clang::Sema::TUFragmentKind)                                                                                                                                        ▒
                           - 27.31% clang::Sema::PerformPendingInstantiations(bool)                                                                                                                                                                 ▒
                              - 27.30% clang::Sema::InstantiateFunctionDefinition(clang::SourceLocation, clang::FunctionDecl*, bool, bool, bool)                                                                                                    ▒
                                 - 23.91% clang::Sema::PerformPendingInstantiations(bool)                                                                                                                                                           ▒
                                    - 23.90% clang::Sema::InstantiateFunctionDefinition(clang::SourceLocation, clang::FunctionDecl*, bool, bool, bool)                                                                                              ▒
                                       - 21.12% clang::Sema::PerformPendingInstantiations(bool)                                                                                                                                                     ▒
                                          - 21.12% clang::Sema::InstantiateFunctionDefinition(clang::SourceLocation, clang::FunctionDecl*, bool, bool, bool)                                                                                        ▒
                                             - 18.12% clang::Sema::PerformPendingInstantiations(bool)                                                                                                                                               ▒
                                                - 18.12% clang::Sema::InstantiateFunctionDefinition(clang::SourceLocation, clang::FunctionDecl*, bool, bool, bool)                                                                                  ▒
                                                   - 16.15% clang::Sema::PerformPendingInstantiations(bool)                                                                                                                                         ▒
                                                      - 16.15% clang::Sema::InstantiateFunctionDefinition(clang::SourceLocation, clang::FunctionDecl*, bool, bool, bool)                                                                            ▒
                                                         - 13.18% clang::Sema::PerformPendingInstantiations(bool)                                                                                                                                   ▒
                                                            - 13.18% clang::Sema::InstantiateFunctionDefinition(clang::SourceLocation, clang::FunctionDecl*, bool, bool, bool)                                                                      ▒                                                               - 11.07% clang::Sema::PerformPendingInstantiations(bool)                                                                                                                             ▒
                                                                  - 11.06% clang::Sema::InstantiateFunctionDefinition(clang::SourceLocation, clang::FunctionDecl*, bool, bool, bool)                                                                ▒
                                                                     + 8.87% clang::Sema::PerformPendingInstantiations(bool)                                                                                                                        ▒
                                                                     + 1.73% clang::Sema::SubstStmt(clang::Stmt*, clang::MultiLevelTemplateArgumentList const&)                                                                                     ▒
                                                               + 1.69% clang::Sema::SubstStmt(clang::Stmt*, clang::MultiLevelTemplateArgumentList const&)                                                                                           ▒
                                                         + 2.38% clang::Sema::SubstStmt(clang::Stmt*, clang::MultiLevelTemplateArgumentList const&)                                                                                                 ▒
                                                   + 1.46% clang::Sema::SubstStmt(clang::Stmt*, clang::MultiLevelTemplateArgumentList const&)                                                                                                       ▒
                                             + 2.11% clang::Sema::SubstStmt(clang::Stmt*, clang::MultiLevelTemplateArgumentList const&)                                                                                                             ▒
                                               0.68% clang::Sema::InstantiateMemInitializers(clang::CXXConstructorDecl*, clang::CXXConstructorDecl const*, clang::MultiLevelTemplateArgumentList const&)                                            ▒
                                       + 2.21% clang::Sema::SubstStmt(clang::Stmt*, clang::MultiLevelTemplateArgumentList const&)                                                                                                                   ▒
                                 + 2.53% clang::Sema::SubstStmt(clang::Stmt*, clang::MultiLevelTemplateArgumentList const&)                                                                                                                         ▒
                                 + 0.72% clang::Sema::InstantiateMemInitializers(clang::CXXConstructorDecl*, clang::CXXConstructorDecl const*, clang::MultiLevelTemplateArgumentList const&)                                                        ▒
                     + 5.44% clang::Preprocessor::Lex(clang::Token&)                                                                                                                                                                                ▒
                  - 7.18% non-virtual thunk to clang::ASTReader::StartTranslationUnit(clang::ASTConsumer*)                                                                                                                                          ▒
                     + clang::ASTReader::PassInterestingDeclsToConsumer()                                                                                                                                                                           ▒
                  + 6.20% clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&)                                                                                                                                                         ▒
            + 2.69% clang::FrontendAction::BeginSourceFile(clang::CompilerInstance&, clang::FrontendInputFile const&)                                                                                                                               ▒
```

Profile after this commit:
```
-   96.85%     0.00%  clang    clang               [.] cc1_main(llvm::ArrayRef<char const*>, char const*, void*)                                                                                                                                    ▒
   - 96.85% cc1_main(llvm::ArrayRef<char const*>, char const*, void*)                                                                                                                                                                               ▒
      - 96.85% clang::ExecuteCompilerInvocation(clang::CompilerInstance*)                                                                                                                                                                           ▒
         - 96.85% clang::CompilerInstance::ExecuteAction(clang::FrontendAction&)                                                                                                                                                                    ▒
            - 96.29% clang::FrontendAction::Execute()                                                                                                                                                                                               ▒
               - 96.29% clang::ParseAST(clang::Sema&, bool, bool)                                                                                                                                                                                   ▒
                  - 79.13% non-virtual thunk to clang::ASTReader::StartTranslationUnit(clang::ASTConsumer*)                                                                                                                                         ▒
                     - clang::ASTReader::PassInterestingDeclsToConsumer()                                                                                                                                                                           ▒
                        - 79.11% clang::BackendConsumer::HandleTopLevelDecl(clang::DeclGroupRef)                                                                                                                                                    ▒
                           - 79.11% (anonymous namespace)::CodeGeneratorImpl::HandleTopLevelDecl(clang::DeclGroupRef)                                                                                                                               ▒
                              - 78.91% (anonymous namespace)::ItaniumCXXABI::EmitCXXConstructors(clang::CXXConstructorDecl const*)                                                                                                                  ▒
                                 - 78.89% clang::Redeclarable<clang::TagDecl>::DeclLink::getPrevious(clang::TagDecl const*) const                                                                                                                   ▒
                                    - clang::ASTReader::CompleteRedeclChain(clang::Decl const*)                                                                                                                                                     ▒
                                       - 78.89% clang::DeclContext::lookup(clang::DeclarationName) const                                                                                                                                            ▒
                                          - 78.89% clang::Redeclarable<clang::NamespaceDecl>::DeclLink::getPrevious(clang::NamespaceDecl const*) const                                                                                              ▒
                                               clang::ASTReader::CompleteRedeclChain(clang::Decl const*)                                                                                                                                            ▒
                                             - clang::DeclContext::lookup(clang::DeclarationName) const                                                                                                                                             ▒
                                                - 78.88% clang::ASTReader::FindExternalVisibleDeclsByName(clang::DeclContext const*, clang::DeclarationName)                                                                                        ▒
                                                   - 78.88% clang::serialization::MultiOnDiskHashTable<clang::serialization::reader::ASTDeclContextNameLookupTrait>::find(clang::DeclarationName const&)                                            ▒
                                                      - 78.83% clang::serialization::MultiOnDiskHashTable<clang::serialization::reader::ASTDeclContextNameLookupTrait>::condense()                                                                  ▒
                                                           21.98% llvm::DenseMap<clang::GlobalDeclID, llvm::detail::DenseSetEmpty, llvm::DenseMapInfo<clang::GlobalDeclID, void>, llvm::detail::DenseSetPair<clang::GlobalDeclID> >::grow(unsigned i▒
                  - 15.74% clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, clang::Sema::ModuleImportState&)                                                                                                                ▒
                     + 8.46% clang::Parser::ParseExternalDeclaration(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*)                                                                                                   ▒
                     + 6.10% clang::Sema::ActOnEndOfTranslationUnit()                                                                                                                                                                               ▒
                     + 1.18% clang::Preprocessor::Lex(clang::Token&)                                                                                                                                                                                ▒
                  + 1.33% clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&)                                                                                                                                                         ▒
            + 0.57% clang::FrontendAction::BeginSourceFile(clang::CompilerInstance&, clang::FrontendInputFile const&)                                                                                                                               ▒
```

alexfh commented:
Spending ~80% of the compilation time in `clang::serialization::MultiOnDiskHashTable` seems wrong.

It also seems like https://github.com/llvm/llvm-project/pull/95348 won't solve this problem, though if you have strong reasons to believe it will, I can try.

As for a reproducer, providing reproducers for modular compilations proved to be an extremely time consuming task. In this particular case we're talking about a compilation that loads more than a thousand dependent modules. It can take quite some time to figure out which ones are contributing most to the clang runtime (`-ftime-report` only shows itemization for `Preloading ....pcm`, which sum up to less than a second, while the full compilation slowed down from ~100s to ~600s).

ilya-biryukov commented:
I added a printf into `FindExternalVisibleDeclsByName` where it calls find on `MultiOnDiskHashTable` to see what names have been causing this and I can see that we stall on the first name we try to look up. The name `std` is the first one and was previously processed almost instantly and now it takes >5 minutes to return results for it. I believe this is where most, if not all, of the overhead is coming from.

Will be digging more, but maybe this gives some clues to where the problem is.

ilya-biryukov commented:
In both cases we are condensing `1105` on-disk hash tables, but it took seconds before and takes multiple minutes after the change.

ilya-biryukov commented:
I only managed to confirm that the problem is that merging of individual PCMs into a merged hash table has become much slower with this change.
I tried with #95348 too, but it didn't help.

I will continue my investigations and reductions tomorrow.

ChuanqiXu9 commented:
Thanks for the profiling data. It narrows the scope a lot. But it makes me confusing too. Since the scope is pretty narrow, we can do an analysis here: https://github.com/llvm/llvm-project/blob/a7a1195f01037e5019f671c96ef4bca9af9bb9a7/clang/lib/Serialization/MultiOnDiskHashTable.h#L145-L173

In `clang::serialization::MultiOnDiskHashTable::condense()`, only https://github.com/llvm/llvm-project/blob/a7a1195f01037e5019f671c96ef4bca9af9bb9a7/clang/lib/Serialization/MultiOnDiskHashTable.h#L160 and https://github.com/llvm/llvm-project/blob/a7a1195f01037e5019f671c96ef4bca9af9bb9a7/clang/lib/Serialization/MultiOnDiskHashTable.h#L163-L164 should be affected by this patch. Where the change in https://github.com/llvm/llvm-project/blob/a7a1195f01037e5019f671c96ef4bca9af9bb9a7/clang/lib/Serialization/MultiOnDiskHashTable.h#L160 should almost be constant since it reads the length.

And for https://github.com/llvm/llvm-project/blob/a7a1195f01037e5019f671c96ef4bca9af9bb9a7/clang/lib/Serialization/MultiOnDiskHashTable.h#L163-L164, its implementation is: https://github.com/llvm/llvm-project/blob/43bd7ae65af40ff3378d5a0395a058ba834ad8dd/clang/lib/Serialization/ASTReader.cpp#L1215-L1225

For `ASTReader::getGlobalDeclID()`, the implementation is https://github.com/llvm/llvm-project/blob/43bd7ae65af40ff3378d5a0395a058ba834ad8dd/clang/lib/Serialization/ASTReader.cpp#L7623-L7644, what is almost bit operation comparing to the previous state.

So it just looks like the reason is that we extend the DeclID from uint32_t to uint64_t and there are too many on-disk hash tables (not a blame). @alexfh could you try to profile on the statements level?

---

BTW, I feel we may read too many times in `MultiOnDiskHashTable `, maybe we can improve this.

ChuanqiXu9 commented:
I sent https://github.com/llvm/llvm-project/pull/95506. It is a independent patch which may mitigate the issue you met. @alexfh you can try this when you have time.

ilya-biryukov commented:
While this looks like a useful optimization, I am still not sure why 2x more data (is that a correct upper bound?) can lead to a 10x slowdown. I will try to dig further to understand what's going on here.

ilya-biryukov commented:
I got to the bottom of it. The problem is that our hash function for 64 bit ints is [not very good](https://github.com/llvm/llvm-project/blob/d5297b72aa32ad3a69563a1fcc61294282f0b379/llvm/include/llvm/ADT/DenseMapInfo.h#L140).

It will have a lot of collision when lower 32 bits are the same and the higher 32 bits of a 64 bit value change. Coincidentally, this is quite common in this case because collisions in low bits of `GlobalDeclID` are very much expected.

I tried replacing it with a variant of MurmurHash I found randomly on the internet and performance went back to normal.
@ChuanqiXu9 I think changing the default hash function for ints might have some far-reaching effects for LLVM code using `DenseSet` and would warrant a bigger discussion on what that function should be (I am not an expert in hash functions myself).
I will be happy to start that discussion on Monday, but feel free to jump into it earlier, if you have time.

In the meantime, I would again propose to revert the change until we can fix the hash function.
(Alternatively, we can change the hash function for `GlobalDeclID` specifically to workaround more locally, but we surely want a better hash function for uint64 in general)

alexfh commented:
Thank you for the analysis, Ilya!

Re: a possible resolution, I wonder why `llvm::hash_value` from llvm/ADT/Hashing.h is not used in `DenseSet`? In any case, if this can't be resolved quickly and needs a design discussion, it would be appropriate to revert the patch first.

ChuanqiXu9 commented:
> I got to the bottom of it. The problem is that our hash function for 64 bit ints is [not very good](https://github.com/llvm/llvm-project/blob/d5297b72aa32ad3a69563a1fcc61294282f0b379/llvm/include/llvm/ADT/DenseMapInfo.h#L140).
> 
> It will have a lot of collision when lower 32 bits are the same and the higher 32 bits of a 64 bit value change. Coincidentally, this is quite common in this case because collisions in low bits of `GlobalDeclID` are very much expected.
> 
> I tried replacing it with a variant of MurmurHash I found randomly on the internet and performance went back to normal. @ChuanqiXu9 I think changing the default hash function for ints might have some far-reaching effects for LLVM code using `DenseSet` and would warrant a bigger discussion on what that function should be (I am not an expert in hash functions myself). I will be happy to start that discussion on Monday, but feel free to jump into it earlier, if you have time.
> 
> In the meantime, I would again propose to revert the change until we can fix the hash function. (Alternatively, we can change the hash function for `GlobalDeclID` specifically to workaround more locally, but we surely want a better hash function for uint64 in general)

Thanks for the analysis. It's pretty helpful. For reverting, on the one hand, we've already targeted the root issue. I think then it is not so rationale to me to block the patch to problems in that level. On the other hand, (I think) the patch (series) is pretty important for named modules (and probably clang explicit modules). And I really want to make that in 19. Given 19 is going to be branched in the end of the next month. From my experience, it is not very hopeful to land such a fundamental change quickly. So I propose to change the hash function in Serialization locally as a workaround and propose a discussion about replacing the hash function in discussion parallelly.

ilya-biryukov commented:
Oh, I didn't realize you were eager to land this in Clang 19, thanks for sharing that.
In that case, I think a specialiazed hash function for `GlobalDeclID` is indeed the way to go.

I was also worried a little there are other performance implications of this change that would block us, but we won't know until we run a full release testing cycle, which may take a week or more. Hopefully it will be okay, as my observations around the increase in PCM sizes align with the numbers from this patch and performance seems to be on par as well, if we change the hash function.
However, I still wanted to mention it so that we won't come out completely of the blue.

I'll try to help reviewing the two patches you posted, thanks!

ChuanqiXu9 commented:
> Oh, I didn't realize you were eager to land this in Clang 19, thanks for sharing that. In that case, I think a specialiazed hash function for `GlobalDeclID` is indeed the way to go.
> 
> I was also worried a little there are other performance implications of this change that would block us, but we won't know until we run a full release testing cycle, which may take a week or more. Hopefully it will be okay, as my observations around the increase in PCM sizes align with the numbers from this patch and performance seems to be on par as well, if we change the hash function. However, I still wanted to mention it so that we won't come out completely of the blue.
> 
> I'll try to help reviewing the two patches you posted, thanks!

Thanks. I've added you as reviewers. I hope to land these 2 patches soon so that we can have more baking time for them.
